From 3f1276e8e286dbd097448acfabab1c1847d04dd2 Mon Sep 17 00:00:00 2001
From: Michal Marek <mmarek@suse.cz>
Date: Wed, 18 Mar 2015 21:34:45 +0100
Subject: [PATCH] kgr: Allow patches to require an exact kernel version
References: bnc#920615
Patch-mainline: No, SLE policy

The idea is that our kgraft patches will set this, to further enforce the
dependencies in the rpm packages.

Signed-off-by: Michal Marek <mmarek@suse.cz>
---
 include/linux/kgraft.h          |    2 ++
 kernel/kgraft.c                 |    6 ++++++
 samples/kgraft/kgraft_patcher.c |    2 ++
 3 files changed, 10 insertions(+)

--- a/include/linux/kgraft.h
+++ b/include/linux/kgraft.h
@@ -78,6 +78,7 @@ struct kgr_patch_fun {
  * @refs: how many patches need to be reverted before this one
  * @name: name of the patch (to appear in sysfs)
  * @owner: module to refcount on patching
+ * @vermagic: VERMAGIC_STRING of the kernel used to build the module
  * @replace_all: revert everything applied before and apply this one instead
  * @immediate: avoid the lazy-switching mechanism and flip the switch ASAP
  * @patches: array of @kgr_patch_fun structures
@@ -92,6 +93,7 @@ struct kgr_patch {
 	/* a patch shall set these */
 	const char *name;
 	struct module *owner;
+	const char *vermagic;
 	bool replace_all;
 	bool immediate;
 	struct kgr_patch_fun patches[];
--- a/kernel/kgraft.c
+++ b/kernel/kgraft.c
@@ -27,6 +27,7 @@
 #include <linux/sort.h>
 #include <linux/spinlock.h>
 #include <linux/types.h>
+#include <linux/vermagic.h>
 #include <linux/workqueue.h>
 
 static int kgr_patch_code(struct kgr_patch_fun *patch_fun, bool final,
@@ -806,6 +807,11 @@ int kgr_patch_kernel(struct kgr_patch *p
 {
 	int ret;
 
+	if (patch->vermagic && strcmp(patch->vermagic, VERMAGIC_STRING) != 0) {
+		pr_err("kgr: kernel version mismatch: \"%s\" != \"%s\"\n",
+				patch->vermagic, VERMAGIC_STRING);
+		return -EINVAL;
+	}
 	if (!try_module_get(patch->owner)) {
 		pr_err("kgr: can't increase patch module refcount\n");
 		return -EBUSY;
--- a/samples/kgraft/kgraft_patcher.c
+++ b/samples/kgraft/kgraft_patcher.c
@@ -24,6 +24,7 @@
 #include <linux/types.h>
 #include <linux/capability.h>
 #include <linux/ptrace.h>
+#include <linux/vermagic.h>
 
 #include <asm/processor.h>
 
@@ -70,6 +71,7 @@ static void kgr_new_function(unsigned lo
 static struct kgr_patch patch = {
 	.name = "sample_patcher",
 	.owner = THIS_MODULE,
+	.vermagic = VERMAGIC_STRING,
 	.patches = {
 #if defined(CONFIG_X86)
 		KGR_PATCH(SyS_iopl, kgr_new_sys_iopl, true),
